### ISSUE TITLE
No error message, FPS unchanged

### ISSUE THREAD

**Issue by king-alp** [Lahvuun/sekirofpsunlock#3]
Followed the readme after downloading sekirofpsunlock-v0.2.0 and ran the command `./sekirofpsunlock 60 set-fps 164` with the game running, the command completed instantly, but the FPS in-game was still locked to 60.
I also attempted to run the command, and then launch the game, which did throw an error message after the game had loaded.
```
./sekirofpsunlock 60 set-fps 164
fgets() failed
is_process_sekiro() failed
find_sekiro_in_dir() failed
find_sekiro() failed
patch() failed
```
I did try running as a normal user and with sudo just in case, my specs are as follows:
Manjaro 21.3.2 KDE 5.95.0 Plasma 5.24.5
Linux Kernel - x86_64 5.15.50-1-MANJARO
AMD Ryzen 5 2600
Nvidia RTX 3070Ti with NVIDIA driver 515.57
GE-Proton7-24, and tried with proton experimental as well.

Is there something I'm doing wrong on my end? Is there a configuration item I'm missing to make it work?

**Comment by Lahvuun** [comment]
Doesn't look like you're doing anything wrong.

You're describing two different problems here:
- FPS lock does not get patched while the game is running. I'm going to need a memory dump of your game to figure out what gets (or doesn't get) patched. I'll think of a simple way you can do it and post it here.
- The patcher looks for the game and exits with errors after failing to read `/proc/<pid>/status` of a process. I have never experienced this, not sure why it might be happening. How reproducible is it? Do you get errors every time you try, or does it sometimes work fine?

**Comment by king-alp** [comment]
Hi! I only tried to run the patcher before launching the game once, but I'll try to replicate that issue now.

On the first attempt it did not throw any error code, but did not unlock the FPS. On a second attempt to run the command before running the game, it caused the same error:

```
./sekirofpsunlock 60 set-fps 164 
fgets() failed
is_process_sekiro() failed
find_sekiro_in_dir() failed
find_sekiro() failed
patch() failed
```
On a third, fourth and fifth attempt the patcher seems to successfully find and then unsuccessfully attempt to patch the game with no error codes. When these errors do occur, the patcher process does not throw these errors until the game opens, it sits and waits until the game launches and then produces those error messages.

Running the patcher while the game is running seems to succeed every time, but does not change the FPS. The patcher does not give any output when I run it with the game running, is that normal operation when a patch is successful? Is there a verbose option to output which memory address is getting patched?

I'm still running GE-Proton7-24, with no launch options, and attempted the above with steam overlay disabled as well to see if it was interfering, netting the same results. I'm not sure how to dump the memory of my game, so pointing me in a way to do so would be nice. I'm new to linux gaming, but not to linux in general, but by far am no linux guru.

As a side note, I did also attempt to patch the FPS with v0.1.0 of this patcher, with the same result, no error message, but no FPS unlock. Thank you for taking the time to look at this issue, and for making this patcher in the first place.

**Comment by Lahvuun** [comment]
Looks like there was a race condition that I was not aware of.
Normally the patcher looks for `sekiro.exe` by repeatedly reading each `/proc/<pid>/status` file. However, it seems that if the process dies after you opened the file but before you read from it, the read will fail. This is surprising (files normally stay around until you close them, even if they are deleted), but makes sense when you think about it — `status` is _not_ a real file.
Anyway, I patched the logic to not fail if that happens, the commit is 60b29c79b422ce75e13ab9be927ca1d6e88dd9c6, you can build it if you want to test it. This should solve the second problem.

To dump the memory I pushed a script at `contrib/dumpsection.py`, you can download it from https://raw.githubusercontent.com/Lahvuun/sekirofpsunlock/ffe74cb10e3e5d64f02efe3ff13d24e6ec1d79ce/contrib/dumpsection.py
Fire up the game, wait until you're in the main menu and then run the script like so:
```sh
./dumpsection.py /proc/$(pgrep sekiro.exe)/mem .text > text
```
This will dump the section of the game's memory that has code in it. You'll get a file called `text`, should be around 40 megabytes in size. You should compress it and upload here. You can compress with tar:
```shell
tar -caf text.tar.gz text
```
It'll take a few seconds. Then, upload `text.tar.gz` here and I'll try to figure out why the FPS lock does not get changed.

**Comment by Lahvuun** [comment]
What's interesting is that I can't set the frame lock higher than 75 (the refresh rate of my only monitor), either. Are you using a 60hz monitor?

**Comment by king-alp** [comment]
Hm, when you try to set the frame lock higher than 75 does it stay at 60? I'm using a 164hz monitor, which is why I used "164" with the patcher, which worked when i played in windows with the patcher that inspired yours. 

Later on I'll attempt to patch at a lower framerate and see if it's able to change it to that. I'm migrating my linux install over to a SSD so I'm unable to test at the moment, so I'll do so after work later tonight when I have steam set back up, and dump the memory as requested as well.

**Comment by Lahvuun** [comment]
> Hm, when you try to set the frame lock higher than 75 does it stay at 60?

No, it goes to 75.

> worked when i played in windows with the patcher that inspired yours

Probably because Windows works in a very different way when it comes to presentation.

> and dump the memory as requested as well

I don't think the problem is with the memory, so no need for the dump for now. See the following section.

---

I was able to reach FPS higher than 75 by using `dxvk.conf` with the following contents:
```
dxgi.syncInterval = 0
```
![2022-07-12 19:11:47 367716609](https://user-images.githubusercontent.com/6643565/178541458-f4970580-fd9d-4fa8-9f61-9d06916140d6.png)
Try putting a file called `dxvk.conf` in the game's root folder (next to sekiro.exe) with `dxgi.syncInterval = 0`, that might fix it for you.

So far my theory is that the patcher works correctly (the chance of it patching the wrong thing are astronomically low), and something in the pipeline is limiting the FPS to 60 for you. Probably DXVK. But could be the Nvidia driver. The Windows unlocker mentions needing to disable VSync in the driver's settings, maybe something similar needs to be done under Linux. Either way, try `dxvk.conf`, if that doesn't work, we'll look into other options.

**Comment by king-alp** [comment]
Adding `dxgi.syncInterval = 0` to `dxvk.conf` fixed the issue, it seems that parameter may have disabled vsync? (It seems that way because it also eliminated some input lag) 

I would try to disable vsync in Nvidia's control panel to see if it nets the same result, but I'm unfamiliar with the NVIDIA X Server Settings application and will need to test that later on. But I'm going to close this issue, thanks again for the assistance, you're doing god's work making linux gaming more worthwhile!

Feature request: **_donate button_**

**Comment by Lahvuun** [comment]
> Adding dxgi.syncInterval = 0 to dxvk.conf fixed the issue, it seems that parameter may have disabled vsync? (It seems that way because it also eliminated some input lag)

It _does_ disable VSync, as stated in the example file: https://github.com/doitsujin/dxvk/blob/master/dxvk.conf
```
# Overrides synchronization interval (Vsync) for presentation.
# Setting this to 0 disables vertical synchronization entirely.
# A positive value 'n' will enable Vsync and repeat the same
# image n times, and a negative value will have no effect.
#
# Supported values: Any non-negative number

# dxgi.syncInterval = -1
# d3d9.presentInterval = -1
```
As I understand, games request the interval through https://docs.microsoft.com/en-us/windows/win32/api/dxgi/nf-dxgi-idxgiswapchain-present, but this option lets you disregard that request and use whichever value you want.

I still wonder why your game is limited to 60 FPS, though. You really shouldn't be needing `dxvk.conf` to fix that. If you're going to continue investigating, my first stop would be Nvidia driver settings.

Also, I remember that if you have multiple monitors, Wine may limit the FPS to the lowest refresh rate among them: https://www.reddit.com/r/wine_gaming/comments/752zo4/my_wine_games_have_a_maximum_of_60_hz_any_way_to/do3t3vl/
You never mentioned if you had any other monitors connected besides the 164Hz one, so I figured I'd let you know.

> Feature request: donate button

Would be nice to have it, I guess, but I don't have a proper way to receive payments here. It'll have to wait until a better day.

edit: I also published [v0.2.1](https://github.com/Lahvuun/sekirofpsunlock/releases/tag/v0.2.1), it should fix
```
./sekirofpsunlock 60 set-fps 164 
fgets() failed
is_process_sekiro() failed
find_sekiro_in_dir() failed
find_sekiro() failed
patch() failed
```

**Comment by Letus252** [comment]
> > Hm, when you try to set the frame lock higher than 75 does it stay at 60?
> 
> No, it goes to 75.
> 
> > worked when i played in windows with the patcher that inspired yours
> 
> Probably because Windows works in a very different way when it comes to presentation.
> 
> > and dump the memory as requested as well
> 
> I don't think the problem is with the memory, so no need for the dump for now. See the following section.
> 
> I was able to reach FPS higher than 75 by using `dxvk.conf` with the following contents:
> 
> ```
> dxgi.syncInterval = 0
> ```
> 
> ![2022-07-12 19:11:47 367716609](https://user-images.githubusercontent.com/6643565/178541458-f4970580-fd9d-4fa8-9f61-9d06916140d6.png) Try putting a file called `dxvk.conf` in the game's root folder (next to sekiro.exe) with `dxgi.syncInterval = 0`, that might fix it for you.
> 
> So far my theory is that the patcher works correctly (the chance of it patching the wrong thing are astronomically low), and something in the pipeline is limiting the FPS to 60 for you. Probably DXVK. But could be the Nvidia driver. The Windows unlocker mentions needing to disable VSync in the driver's settings, maybe something similar needs to be done under Linux. Either way, try `dxvk.conf`, if that doesn't work, we'll look into other options.

Hey, just wanted to say; I had similar issue, adding `dxvk.conf` fixed it, thanks.
